### PR TITLE
Make Duration constructors and accessors const fns

### LINKED ISSUE #33029: Make duration constructor functions `const`.
Is there anything stopping the functions `from_millis, from_secs and new` in `std::time::Duration` from being constant?

### PR DESCRIPTION
**Pull request by sfackler**
Closes #33029 

r? @alexcrichton

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Thanks!

r=me

**Comment by alexcrichton** [comment]
@bors: r-

@GuillaumeGomez this is a change which has ramifications on the stability of these APIs and how we might change them in the future, I'd like to discuss this at a libs triage meeting before we move forward with this just yet.

We currently don't have many conventions for what is or what isn't a `const fn` in the standard library just yet, and we have many possible candidates for doing so, so I want to just make sure we approach this in a principled fashion rather than "let's make everything `const` which can be so right now"

**Comment by GuillaumeGomez** [comment]
@alexcrichton: My bad, I didn't think that the change would be that big. I'll wait for your come back on this.

**Comment by steveklabnik** [comment]
@luqmania to be clear, @guillaumegomez doesn't quite have r+ rights yet. I've been working with him on smaller things, but I'm supposed to approve them before they get put in. He forgot to CC me on this one.

I'm surprised Bors allowed this, actually.

**Comment by luqmana** [comment]
@steveklabnik Ah whoops, my mistake. Actually, I don't think bors cares what comes after r= as long as the person actually writing the comment is allowed to.

File: src/libstd/time/duration.rs [hunk 1]
```diff
@@ -81,7 +83,7 @@ impl Duration {
     /// nanoseconds are not represented in the returned value).
     #[stable(feature = "duration", since = "1.3.0")]
     #[inline]
-    pub fn as_secs(&self) -> u64 { self.secs }
+    pub const fn as_secs(&self) -> u64 { self.secs }
```
Comment (oli-obk) [inline src/libstd/time/duration.rs]: These kind of const fns only work in statics, they don't work in array lengths and other true constants (yet). Is there a use case for this? If not, I suggest to delay this until we have a unified MIR-powered const evaluator

**Comment by alexcrichton** [comment]
The libs team discussed this PR during triage today and the decision was to close for now. Adding a `const` fn is a serious API commitment that we may not always be able to uphold (e.g. #33072). In the meantime `const fn` is still unstable, and when stabilized we will likely want to perform a pass over the standard library to audit for functions which are const-compatible at that time.

**Pull request closed without merging**